### PR TITLE
Add types support for Sparse Initializer in Onnxruntime

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Add support for types such as double, float16, bfloat16 and all integers as real models require them.
Fix edge case when sparse initializers do not have any NNZ.

**Motivation and Context**
Support customer use of sparsity.

### CODE REVIEW

File: onnxruntime/test/framework/sparse_kernels_test.cc [hunk 1]
```diff
@@ -511,12 +511,15 @@ std::vector<std::string> CreateValues<std::string>() {
...
 */
 
-/* BFloat16 support in the future
 template <>
 std::vector<BFloat16> CreateValues<BFloat16>() {
   return {BFloat16(1.f), BFloat16(2.f), BFloat16(3.f), BFloat16(4.f)};
 }
-*/
+
+template <>
```
Comment (yuslepukhin) [inline onnxruntime/test/framework/sparse_kernels_test.cc:519]: >emplate  [](start = 1, length = 8)

Add all zeros test case

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 2]
```diff
@@ -1092,10 +1137,66 @@ common::Status DenseTensorToSparseTensorProto(const ONNX_NAMESPACE::TensorProto&
...
+          break;
+        }
+        case TensorProto_DataType_INT16:
+        case TensorProto_DataType_UINT16: {
+          SparsifyGeneric(dense_raw_data.get(), n_dense_elements, element_size,
+                          IsZero<uint16_t>, CopyElement<uint16_t>, values, indices);
+          break;
+        }
+        default:
+          ORT_THROW(false, "BUG! Report to onnxruntime team.");
```
Comment (skottmckay) [inline onnxruntime/core/framework/tensorprotoutils.cc:1160]: >ORT_THROW [](start = 10, length = 9)

More graceful to use ORT_RETURN_IF in a build with exceptions disabled (returns Status vs. abort() being called).

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 3]
```diff
@@ -1092,10 +1137,66 @@ common::Status DenseTensorToSparseTensorProto(const ONNX_NAMESPACE::TensorProto&
...
+        case TensorProto_DataType_INT32:
+        case TensorProto_DataType_UINT32: {
+          SparsifyGeneric(dense_raw_data.get(), n_dense_elements, element_size,
+                          IsZero<uint32_t>, CopyElement<uint32_t>, values, indices);
+          break;
+        }
+        default:
+          ORT_THROW(false, "BUG! Report to onnxruntime team.");
+          break;
+      }
```
Comment (skottmckay) [inline onnxruntime/core/framework/tensorprotoutils.cc:1180]: I understand the MLFloat16 and BFloat16 need explicit special handling, but for anything else would it be good enough just to call SparsifyGeneric with any type that has the correct data size so we don't need these additional 'default:' cases for an unhandled type? We've already explicitly checked the type is in a known list via the previous usage of MLTypeCallDispatcher

File: tools/python/sparsify_initializers.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input', required=True, type=str, help='input model path')
+    parser.add_argument('--output', required=True, type=str, help='output model path')
+    parser.add_argument('--exclude', required=False, type=str,
+                        help='semicolon separated list of initializer names to exclude')
+    parser.add_argument('--tolerance', required=False, type=float, default=1e-6,
+                        help='FP tolerance. If not given simple compare to 0')
```
Comment (skottmckay) [inline tools/python/sparsify_initializers.py:29]: Relative or absolute? Or should you have both?

File: tools/python/sparsify_initializers.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# This script opens an existing model in onnx format and attemptes to
```
Comment (pranavsharma) [inline tools/python/sparsify_initializers.py:5]:
```suggestion
# This script opens an existing model in onnx format and attempts to
```

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 6]
```diff
@@ -966,36 +967,60 @@ common::Status SparseTensorProtoToDenseTensorProto(const ONNX_NAMESPACE::SparseT
...
+              n_sparse_elements,
+              indices, dims,
+              [sparse_data, dense_data_span](size_t from_idx, size_t to_idx) {
+                dense_data_span[to_idx] = static_cast<const uint64_t*>(sparse_data)[from_idx];
+              });
+          break;
+        }
+
+        default:
+          ORT_THROW(false, "BUG! Report to onnxruntime team.");
```
Comment (pranavsharma) [inline onnxruntime/core/framework/tensorprotoutils.cc:1019]: Return status?

File: tools/python/sparsify_initializers.py (hunk 4 above)
Comment (yuslepukhin) [reply tools/python/sparsify_initializers.py:29]: Relative to what? We just need to compare to zero.

---
In reply to: [595635191](https://github.com/microsoft/onnxruntime/pull/7004#discussion_r595635191) [](ancestors = 595635191)

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 7]
```diff
@@ -1081,25 +1108,38 @@ common::Status DenseTensorToSparseTensorProto(const ONNX_NAMESPACE::TensorProto&
   std::unique_ptr<uint8_t[]> dense_raw_data;
   ORT_RETURN_IF_ERROR(UnpackInitializerData(dense_proto, model_path, dense_raw_data, tensor_bytes_size));
   size_t element_size = 0;
-  MLTypeCallDispatcher<float, int8_t, uint8_t> type_disp(data_type);
+  // We want this type list to match the one above in SparseTensorProtoToDenseTensorProto
```
Comment (edgchen1) [inline onnxruntime/core/framework/tensorprotoutils.cc:1111]: regarding the comment about using a matching type list - can do this too:
```cpp
using L = onnxruntime::TypeList<float, double, ...>; // replace L with a better name
// ...
MLTypeCallDispatcherFromTypeList<L> type_disp(data_type);
```

**Review by pranavsharma — DISMISSED** [review]

File: tools/python/sparsify_initializers.py (hunk 4 above)
Comment (yuslepukhin) [reply tools/python/sparsify_initializers.py]: We don't need relative.

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 8]
```diff
@@ -956,51 +960,77 @@ common::Status SparseTensorProtoToDenseTensorProto(const ONNX_NAMESPACE::SparseT
...
+              n_sparse_elements,
+              indices, dims,
+              [sparse_data, dense_data_span](size_t from_idx, size_t to_idx) {
+                dense_data_span[to_idx] = static_cast<const uint64_t*>(sparse_data)[from_idx];
+              });
+          break;
+        }
+
+        default:
+          return ::onnxruntime::common::Status(::onnxruntime::common::ONNXRUNTIME,
```
Comment (skottmckay) [inline onnxruntime/core/framework/tensorprotoutils.cc:1022]: >return ::onnxruntime::common::Status [](start = 10, length = 36)

FWIW could use ORT_MAKE_STATUS(...) or ORT_RETURN_IF(false, "BUG...")

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**